pkg/crd/flatten.go

```diff
@@ -26,47 +27,176 @@ import (
+	hoisted := false
+
+	for i := 0; i < srcVal.NumField(); i++ {
+		fieldName := typ.Field(i).Name
+		if fieldName == "AllOf" {
+			continue
+		}
+		srcField := srcVal.Field(i)
+		fldTyp := srcField.Type()
+		zeroVal := reflect.Zero(fldTyp)
+		zeroInt := zeroVal.Interface()
+		srcInt := srcField.Interface()
+
+		if isOrNil(srcField, srcInt, zeroInt) {
+			// nothing to copy from src, continue
+			continue
+		}
+
+		dstField := dstVal.Field(i)
+		dstInt := dstField.Interface()
+		if isOrNil(dstField, dstInt, zeroInt) {
+			// dst is empty, continue
+			dstField.Set(srcField)
+			continue
+		}
+
+		if fldTyp.Comparable() && srcInt == dstInt {
+			// same value, continue
+			continue
+		}
+
+		// resolve conflict
+		switch fieldName {
+		case "Properties":
+			// merge if possible, use all of otherwise
+			srcMap := srcInt.(map[string]apiext.JSONSchemaProps)
+			dstMap := dstInt.(map[string]apiext.JSONSchemaProps)
+
+			for k, v := range srcMap {
+				dstProp, exists := dstMap[k]
+				if !exists {
+					dstMap[k] = v
+					continue
+				}
+				flattenAllOfInto(&dstProp, v, errRec)
+				dstMap[k] = dstProp
+			}
+		case "Required":
+			// merge
+			dstField.Set(reflect.AppendSlice(dstField, srcField))
+		case "Type":
+			if srcInt != dstInt {
+				// TODO(directxman12): figure out how to attach this back to a useful point in the Go source or in the schema
+				errRec.AddError(fmt.Errorf("conflicting types in allOf branches in schema: %s vs %s", dstInt, srcInt))
+			}
+			// keep the destination value, for now
+		// TODO(directxman12): Default -- use field?
+		// TODO(directxman12):
+		// - Dependencies: if field x is present, then either schema validates or all props are present
+		// - AdditionalItems: like AdditionalProperties
+		// - Definitions: common named validation sets that can be references (merge, bail if duplicate)
+		case "AdditionalProperties":
+			// TODO(directxman12): handle the bool too
+			srcProps := srcInt.(*apiext.JSONSchemaPropsOrBool)
+			if srcProps.Schema == nil {
+				// nothing to merge
+				continue
+			}
+			dstProps := dstInt.(*apiext.JSONSchemaPropsOrBool)
+			if dstProps.Schema == nil {
+				dstProps.Schema = &apiext.JSONSchemaProps{}
+			}
+			flattenAllOfInto(dstProps.Schema, *srcProps.Schema, errRec)
+		case "Nullable":
+			// true overrides false
+			if srcInt.(bool) {
+				dstField.Set(srcField)
+			}
+		// TODO(directxman12): src isn't necessarily the field value -- it's just the most recent allOf entry
+		case "Title", "Description", "Example", "ExternalDocs":
+			// merged before, don't touch
+		default:
+			// hoist into allOf...
+			hoisted = true
+
+			srcRemVal.Field(i).Set(srcField)
+			dstRemVal.Field(i).Set(dstField)
+			// ...and clear the original
+			dstField.Set(zeroVal)
+		}
 	}
-	for _, propName := range src.Required {
-		dst.Required = append(dst.Required, propName)
+
+	if hoisted {
+		dst.AllOf = append(dst.AllOf, dstRemainder, srcRemainder)
```

Comment:
this will potentially not return a structural schema. Types, nullable, etc., i.e. everything from https://github.com/kubernetes/kubernetes/blob/9ecc57ef880c475c052e4ab472307202499cf8d9/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/structural.go#L47 has to be pushed into the root. Those cannot be under `allOf`.

Reply:
right -- that's what the above does -- we only `hoist` for properties that aren't listed above, so `type`, `nullable`, etc should be handled there.

Reply:
Towards structural schema, take a look at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go#L63. You might want to call that somewhere, or even use the Structural type directly.

Reply:
let's file a follow-up issue for that.